tofd/pars/R3BTofDCal2HitPar.cxx

```diff
@@ -297,1099 +352,2668 @@ void R3BTofDCal2HitPar::Exec(Option_t* option)
+
+			    qb = qb * fTofdQ;
+			    }*/
+		
+			}
+		    // prepare double exponential fit
+		    if ((!fTofdSmiley && fTofdQ > 0.1 && fParameter > 1 && fParameter != 10))
+		    {
+				LOG(debug) << "Prepare histo for double exponential fit";
+			    auto par = fHitPar->GetModuleParAt(iPlane, iBar);
+			    if (!par)
+			    {
+				    R3BLOG(error,
+						    "Hit par not found, Plane: " << top->GetDetectorId() << ", Bar: " << top->GetBarId());
+				    continue;
+			    }
+				//Ctop_tot = top_tot  * par->GetToTOffset3() - par->GetToTOffset4();
+				//Cbot_tot = bot_tot  * par->GetToTOffset1() - par->GetToTOffset2();
+				Ctop_tot = top_tot  * par->GetToTOffset2();
+				Cbot_tot = bot_tot  * par->GetToTOffset1();
+
+			    // calculate y position
+			    auto pos = ((bot_ns + par->GetOffset1()) - (top_ns + par->GetOffset2())) * par->GetVeff();
+
+			    // fill fitting histograms and smiley histogram
+			    fhTot1vsPos[iPlane - 1][iBar - 1]->Fill(pos, Cbot_tot);
+			    fhTot2vsPos[iPlane - 1][iBar - 1]->Fill(pos, Ctop_tot);
+
+			}
+
+		    // std::cout<< " hola 10" << fTofdZ << std::endl;
+
+		    // prepare charge fit / quench correction
+		    if (fTofdZ == true && fParameter > 1 && fParameter != 10)
+		    {
+				
+			    // std::cout<< " hola 20" << std::endl;
+			    LOG(debug) << "Prepare histo for quenching correction";
+			    // get parameter
+			    auto par = fHitPar->GetModuleParAt(iPlane, iBar);
+			    if (!par)
+			    {
+				    R3BLOG(error,
+						    "Hit par not found, Plane: " << top->GetDetectorId() << ", Bar: " << top->GetBarId());
+				    continue;
+			    }
+
+			    // Calculate y position
+			    auto posToT =
+				    par->GetLambda() * log((Ctop_tot) / (Cbot_tot));
+
+
+			    Double_t parf[4];
+			    parf[0] = par->GetPar1a();
+			    parf[1] = par->GetPar1b();
+			    parf[2] = par->GetPar1c();
+			    parf[3] = par->GetPar1d();
+
+
+			    Double_t parq[4];
+			    parq[0] = par->GetPola();
+			    parq[1] = par->GetPolb();
+			    parq[2] = par->GetPolc();
+			    parq[3] = par->GetPold();
+
+			    // Calculate charge
+			    Double_t qb = 0.;
+			    Double_t qbx = 0.;
+			    if (fTofdSmiley)
+			    {	
+					if(parq[0]==0. && parq[1]==0. && parq[2]==0. && parq[3]==0.) parq[0] = 1.;
+
+				    qb = TMath::Sqrt(Ctop_tot * Cbot_tot);///(parq[0] + parq[1] * posToT + parq[2] * pow(posToT, 2) + parq[3] * pow(posToT, 3));
+				    qbx = TMath::Sqrt(Ctop_tot * Cbot_tot)*(parq[0] + parq[1] * posToT + parq[2] * pow(posToT, 2) + parq[3] * pow(posToT, 3));
+				    qb = qb *
+					    fTofdQ / 200.; // theory says: dE ~ Z^2 but we see quenching -> just use linear and fit the rest
+					qbx = qbx * fTofdQ / 200.;
+				     std::cout<< " hola30 " << qb << " , " << fTofdQ << " , "<<qbx<<std::endl;
+					 if(qbx > 70.){
+						qbx = qbx / 200.;
+						cout<<"XXXXXXX "<< qb << " , " << fTofdQ << " , "<<qbx<<std::endl;
+					 }
```

Comment:
200 is the ToT value consistent with where the beam was. It's a hard-coded value - probably not a good ideal

The code effectively says take the sqrt(ToT1*ToT2) (should be approx. 200), multiply it by 8 (charge of the O-beam at ToT=200) and dividing by the value it should be. Effectively, this is a charge calibration at 1 point.

Agree, it should not be hard-coded, but set as a variable.

Reply:
Ok I agree for `qb` here. If there are no charge calibration parameters this is a good first guess.

But isn't the function that gives this parq parameters in line 688 fitted with just `sqrt(ToT1*ToT2)` as y-values filled in lines 514 or 596? With a pol3 fit this should describe the smiley/frowny/whatever, at the ToT that was aimed for in the calibration beam?

In the classical case I expect `sqrt(ToT1*ToT2)` to rise on the ends of a paddle. 
If I muliply the `sqrt(ToT1*ToT2)` (say 210) at this given `posToT` with the value of the function at this point (ideal case also 210 since we only have calibration beam and no fragments) and 8, i end up with something large. That I divide by 200 which is still large (`>70.`) and because of that I divide again by 200 which is then in the range were i want it to be but not 8 (210*210*8/200/200=8.82) and even further away from 8 than the qb method you described (210*8/200=8.4) which doesn't use the fit function??

Before, as commented in line 687 we would've had with this example: 210/210*8=8 were the second 210 is the value of the function.

But maybe I'm just slow today and missed something in the code or the function definition changed...

Anyway, I also don't see the benefit of remembering what value the ToT was set to in the calibration beam each time one wants to run the Cal2Hit macro, why not use the info from the fit and in the macro only the charge?

Reply:
The division is applied only after the smiley corrected value of qb is calculated, so it should not diverge on the edges of the bar anymore. But maybe I now am missing something myself... Maybe this is best further discussed in person.:)